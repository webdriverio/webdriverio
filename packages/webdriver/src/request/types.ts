import type { Options } from '@wdio/types'

export interface WebDriverResponse {
    value: any
    /**
     * error case
     * https://w3c.github.io/webdriver/webdriver-spec.html#dfn-send-an-error
     */
    error?: string
    message?: string
    stacktrace?: string

    /**
     * JSONWP property
     */
    status?: number
    sessionId?: string
}

export type RequestLibResponse = Options.RequestLibResponse
export type RequestOptions = Omit<Options.WebDriver, 'capabilities'>
