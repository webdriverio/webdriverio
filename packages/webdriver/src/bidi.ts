import { EventEmitter } from 'node:events'
import WebSocket from 'ws'
import logger from '@wdio/logger'
import { BidiRequest, BidiResponse } from '@wdio/protocols'

const log = logger('webdriver:BidiHandler')
const RESPONSE_TIMEOUT = 1000 * 60

export class BidiHandler extends EventEmitter {
    private _id = 0
    private _ws: WebSocket
    private _isConnected = false

    constructor (private _webSocketUrl: string) {
        super()
        log.info(`Connect to webSocketUrl ${this._webSocketUrl}`)
        this._ws = new WebSocket(this._webSocketUrl)
    }

    public connect () {
        return new Promise<void>((resolve) => this._ws.on('open', () => {
            this._isConnected = true
            resolve()
        }))
    }

    get socket () {
        return this._ws
    }

    public send (params: BidiRequest) {
        if (!this._isConnected) {
            throw new Error('No connection to WebDriver Bidi was established')
        }

        const id = ++this._id
        this._ws.send(JSON.stringify({ id, ...params }))
        return new Promise<BidiResponse>((resolve, reject) => {
            const t = setTimeout(() => {
                reject(new Error(`Request with id ${id} timed out`))
                h.off('message', listener)
            }, RESPONSE_TIMEOUT)

            const listener = (data: WebSocket.RawData) => {
                try {
                    const payload = JSON.parse(data.toString()) as BidiResponse
                    if (payload.id === id) {
                        clearTimeout(t)
                        h.off('message', listener)
                        resolve(payload)
                    }
                } catch (err: any) {
                    log.error(`Failed parse message: ${err.message}`)
                }
            }
            const h = this._ws.on('message', listener)
        })
    }

    public sendAsync (params: BidiRequest) {
        if (!this._isConnected) {
            throw new Error('No connection to WebDriver Bidi was established')
        }

        const id = ++this._id
        this._ws.send(JSON.stringify({ id, ...params }))
    }
}
