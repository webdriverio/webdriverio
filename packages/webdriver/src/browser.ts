import WebDriver from './index.js'
import { BrowserSocket } from './bidi/socket.js'
import { FetchRequest } from './request/web.js'

export default WebDriver
export * from './index.js'

import { environment } from './environment.js'

environment.value = {
    Request: FetchRequest,
    Socket: BrowserSocket,
    variables: {}
}

