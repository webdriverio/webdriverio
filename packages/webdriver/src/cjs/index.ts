function command (method: string, encodeUri: string, commandInfo: any, doubleEncodeVariables = false) {
    return async function protocolCommand(this: unknown, ...args: unknown[]) {
        const commandESM = await import('../command.js')
        return commandESM.default(method, encodeUri, commandInfo, doubleEncodeVariables).apply(this, args)
    }
}

class WebDriver {
    static async newSession(
        options: any,
        modifier?: (...args: any[]) => any,
        userPrototype = {},
        customCommandWrapper?: (...args: any[]) => any
    ): Promise<any> {
        const WebDriver = (await import('../index.js')).default
        return WebDriver.newSession(options, modifier, userPrototype, customCommandWrapper)
    }

    /**
     * allows user to attach to existing sessions
     */
    static async attachToSession (
        options?: any,
        modifier?: (...args: any[]) => any,
        userPrototype = {},
        commandWrapper?: (...args: any[]) => any
    ): Promise<any> {
        const WebDriver = (await import('../index.js')).default
        return WebDriver.attachToSession(options, modifier, userPrototype, commandWrapper)
    }

    /**
     * Changes The instance session id and browser capabilities for the new session
     * directly into the passed in browser object
     *
     * @param   {object} instance  the object we get from a new browser session.
     * @returns {string}           the new session id of the browser
     */
    static async reloadSession (instance: any) {
        const WebDriver = (await import('../index.js')).default
        return WebDriver.reloadSession(instance)
    }

    static get WebDriver () {
        return WebDriver
    }

    static get command () {
        return command
    }
}

module.exports = WebDriver
exports.command = command
