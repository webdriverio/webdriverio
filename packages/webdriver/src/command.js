import logger from '@wdio/logger'
import WebDriverRequest from './request'
import { isValidParameter, getArgumentType, commandCallStructure } from './utils'

const log = logger('webdriver')

export default function (method, endpointUri, commandInfo) {
    const { command, ref, parameters, variables = [] } = commandInfo

    return function (...args) {
        let endpoint = endpointUri // clone endpointUri in case we change it
        const commandParams = [...variables.map((v) => Object.assign(v, {
            /**
             * url variables are:
             */
            required: true, // always required as they are part of the endpoint
            type: 'string'  // have to be always type of string
        })), ...parameters]

        const commandUsage = `${command}(${commandParams.map((p) => p.name).join(', ')})`
        const moreInfo = `\n\nFor more info see ${ref}\n`
        const body = {}

        /**
         * parameter check
         */
        const minAllowedParams = commandParams.filter((param) => param.required).length
        if (args.length < minAllowedParams || args.length > commandParams.length) {
            const parameterDescription = commandParams.length
                ? `\n\nProperty Description:\n${commandParams.map((p) => `  "${p.name}" (${p.type}): ${p.description}`).join('\n')}`
                : ''

            throw new Error(
                `Wrong parameters applied for ${command}\n` +
                `Usage: ${commandUsage}` +
                parameterDescription +
                moreInfo
            )
        }

        /**
         * parameter type check
         */
        for (const [i, arg] of Object.entries(args)) {
            const commandParam = commandParams[i]

            if (!isValidParameter(arg, commandParam.type)) {
                /**
                 * ignore if argument is not required
                 */
                if (typeof arg === 'undefined' && !commandParam.required) {
                    continue
                }

                throw new Error(
                    `Malformed type for "${commandParam.name}" parameter of command ${command}\n` +
                    `Expected: ${commandParam.type}\n` +
                    `Actual: ${getArgumentType(arg)}` +
                    moreInfo
                )
            }

            /**
             * inject url variables
             */
            if (i < variables.length) {
                endpoint = endpoint.replace(`:${commandParams[i].name}`, arg)
                continue
            }

            /**
             * rest of args are part of body payload
             */
            body[commandParams[i].name] = arg
        }

        const request = new WebDriverRequest(method, endpoint, body)
        this.emit('command', { method, endpoint, body })
        log.info('COMMAND', commandCallStructure(command, args))
        return request.makeRequest(this.options, this.sessionId).then((result) => {
            if (result.value != null) {
                log.info('RESULT', command.toLowerCase().includes('screenshot')
                    && typeof result.value === 'string' && result.value.length > 64
                    ? `${result.value.substr(0, 61)}...` : result.value)
            }

            this.emit('result', { method, endpoint, body, result })
            return result.value
        })
    }
}
