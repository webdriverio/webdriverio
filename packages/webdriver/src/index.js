import logger from '@wdio/logger'
import merge from 'lodash.merge'
import { validateConfig } from '@wdio/config'

import webdriverMonad from './monad'
import WebDriverRequest from './request'
import { DEFAULTS } from './constants'
import { getPrototype, environmentDetector } from './utils'

import WebDriverProtocol from '../protocol/webdriver.json'
import JsonWProtocol from '../protocol/jsonwp.json'
import MJsonWProtocol from '../protocol/mjsonwp.json'
import AppiumProtocol from '../protocol/appium.json'
import ChromiumProtocol from '../protocol/chromium.json'

export default class WebDriver {
    static async newSession (options = {}, modifier, userPrototype = {}, commandWrapper) {
        const params = validateConfig(DEFAULTS, options)
        logger.setLevel('webdriver', params.logLevel)

        /**
         * the user could have passed in either w3c style or jsonwp style caps
         * and we want to pass both styles to the server, which means we need
         * to check what style the user sent in so we know how to construct the
         * object for the other style
         */
        const [w3cCaps, jsonwpCaps] = params.capabilities && params.capabilities.alwaysMatch
            /**
             * in case W3C compliant capabilities are provided
             */
            ? [params.capabilities, params.capabilities.alwaysMatch]
            /**
             * otherwise assume they passed in jsonwp-style caps (flat object)
             */
            : [{ alwaysMatch: params.capabilities, firstMatch: [{}] }, params.capabilities]

        const sessionRequest = new WebDriverRequest(
            'POST',
            '/session',
            {
                capabilities: w3cCaps, // W3C compliant
                desiredCapabilities: jsonwpCaps // JSONWP compliant
            }
        )

        const response = await sessionRequest.makeRequest(params)

        /**
         * save original set of capabilities to allow to request the same session again
         * (e.g. for reloadSession command in WebdriverIO)
         */
        params.requestedCapabilities = { w3cCaps, jsonwpCaps }

        /**
         * save actual receveived session details
         */
        params.capabilities = response.value.capabilities || response.value

        /**
         * apply mobile flags to driver scope
         */
        const { isW3C, isMobile, isIOS, isAndroid, isChrome } = environmentDetector(params.capabilities)
        const environmentFlags = {
            isW3C: { value: isW3C },
            isMobile: { value: isMobile },
            isIOS: { value: isIOS },
            isAndroid: { value: isAndroid },
            isChrome: { value: isChrome }
        }

        const protocolCommands = getPrototype({ isW3C, isMobile, isIOS, isAndroid, isChrome })
        const prototype = merge(protocolCommands, environmentFlags, userPrototype)
        const monad = webdriverMonad(params, modifier, prototype)
        return monad(response.value.sessionId || response.sessionId, commandWrapper)
    }

    /**
     * allows user to attach to existing sessions
     */
    static attachToSession (options = {}, modifier, userPrototype = {}, commandWrapper) {
        if (typeof options.sessionId !== 'string') {
            throw new Error('sessionId is required to attach to existing session')
        }

        logger.setLevel('webdriver', options.logLevel)

        options.capabilities = options.capabilities || {}
        options.isW3C = options.isW3C || true
        const prototype = Object.assign(getPrototype(options.isW3C), userPrototype)
        const monad = webdriverMonad(options, modifier, prototype)
        return monad(options.sessionId, commandWrapper)
    }

    static get WebDriver () {
        return WebDriver
    }
    static get DEFAULTS () {
        return DEFAULTS
    }

    /**
     * Protocols
     */
    static get WebDriverProtocol () {
        return WebDriverProtocol
    }
    static get JsonWProtocol () {
        return JsonWProtocol
    }
    static get MJsonWProtocol () {
        return MJsonWProtocol
    }
    static get AppiumProtocol () {
        return AppiumProtocol
    }
    static get ChromiumProtocol () {
        return ChromiumProtocol
    }
}

/**
 * Helper methods consumed by webdriverio package
 */
export {
    webdriverMonad,
    getPrototype
}
