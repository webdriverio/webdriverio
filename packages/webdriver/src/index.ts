import path from 'path'
import logger from '@wdio/logger'

import { webdriverMonad, sessionEnvironmentDetector } from '@wdio/utils'
import { validateConfig } from '@wdio/config'

import { DEFAULTS } from './constants'
import { startWebDriverSession, getPrototype, getEnvironmentVars } from './utils'
import type {
    Options, Client, AttachOptions, SessionFlags,
    DesiredCapabilities
} from './types'

const log = logger('webdriver')

export default class WebDriver {
    static async newSession (
        options: Options = {},
        modifier?: (...args: any[]) => any,
        userPrototype = {},
        customCommandWrapper?: (...args: any[]) => any
    ): Promise<Client> {
        const params = validateConfig(DEFAULTS, options)

        if (!options.logLevels || !options.logLevels.webdriver) {
            logger.setLevel('webdriver', params.logLevel!)
        }

        /**
         * if the server responded with direct connect information, update the
         * params to speak directly to the appium host instead of a load
         * balancer (see https://github.com/appium/python-client#direct-connect-urls
         * for example). But only do this if the user has enabled this
         * behavior in the first place.
         */
        const { directConnectProtocol, directConnectHost, directConnectPort, directConnectPath } = params
        if (directConnectProtocol && directConnectHost && directConnectPort && (directConnectPath || directConnectPath === '')) {
            log.info('Found direct connect information in new session response. ' +
                `Will connect to server at ${directConnectProtocol}://${directConnectHost}:${directConnectPort}/${directConnectPath}`)
            params.protocol = directConnectProtocol
            params.hostname = directConnectHost
            params.port = directConnectPort
            params.path = directConnectPath
        }

        /**
         * Store all log events in a file
         */
        if (params.outputDir) {
            process.env.WDIO_LOG_PATH = path.join(params.outputDir, 'wdio.log')
        }

        const { sessionId, capabilities } = await startWebDriverSession(params)
        const environment = sessionEnvironmentDetector({ capabilities, requestedCapabilities: params.requestedCapabilities })
        const environmentPrototype = getEnvironmentVars(environment)
        const protocolCommands = getPrototype(environment)
        const prototype = { ...protocolCommands, ...environmentPrototype, ...userPrototype }

        const monad = webdriverMonad(params, modifier, prototype)
        return monad(sessionId, customCommandWrapper)
    }

    /**
     * allows user to attach to existing sessions
     */
    static attachToSession (
        options?: AttachOptions,
        modifier?: (...args: any[]) => any,
        userPrototype = {},
        commandWrapper?: (...args: any[]) => any
    ): Client {
        if (!options || typeof options.sessionId !== 'string') {
            throw new Error('sessionId is required to attach to existing session')
        }

        // logLevel can be undefined in watch mode when SIGINT is called
        if (options.logLevel !== undefined) {
            logger.setLevel('webdriver', options.logLevel)
        }

        options.capabilities = options.capabilities || {}
        options.isW3C = options.isW3C === false ? false : true

        const environmentPrototype = getEnvironmentVars(options as Partial<SessionFlags>)
        const protocolCommands = getPrototype(options as Partial<SessionFlags>)
        const prototype = { ...protocolCommands, ...environmentPrototype, ...userPrototype }
        const monad = webdriverMonad(options, modifier, prototype)
        return monad(options.sessionId, commandWrapper)
    }

    /**
     * Changes The instance session id and browser capabilities for the new session
     * directly into the passed in browser object
     *
     * @param   {Object} instance  the object we get from a new browser session.
     * @returns {string}           the new session id of the browser
    */
    static async reloadSession (instance: Client) {
        const params: Options = {
            ...instance.options,
            capabilities: instance.requestedCapabilities as DesiredCapabilities
        }
        const { sessionId, capabilities } = await startWebDriverSession(params)
        instance.sessionId = sessionId
        instance.capabilities = capabilities
        return sessionId
    }

    static get WebDriver () {
        return WebDriver
    }
}

/**
 * Helper methods consumed by webdriverio package
 */
export { getPrototype, DEFAULTS }
