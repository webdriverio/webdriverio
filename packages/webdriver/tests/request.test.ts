import logger from '@wdio/logger'
import * as gotOrig from 'got'
import http from 'http'
import https from 'https'

import WebDriverRequest from '../src/request'
import { WebDriverLogTypes } from '../src/types'

type LogMock = {
    warn: jest.Mock,
    error: jest.Mock
}

const { warn, error } = logger('test') as unknown as LogMock
const got = gotOrig.default as unknown as jest.Mock

const path = '/session'
const defaultOptions = {
    protocol: 'http',
    hostname: 'localhost',
    port: 4444
}

describe('webdriver request', () => {
    beforeEach(() => {
        got.mockClear()
    })

    it('should have some default options', () => {
        const req = new WebDriverRequest('POST', '/foo/bar', { foo: 'bar' })
        expect(req.method).toBe('POST')
        expect(req.endpoint).toBe('/foo/bar')
    })

    it('should be able to make request', () => {
        const req = new WebDriverRequest('POST', '/foo/bar', { foo: 'bar' })
        req['_createOptions'] = jest.fn().mockImplementation((opts, sessionId) => ({
            foo: 'bar',
            sessionId
        }))
        req.emit = jest.fn()
        req['_request'] = jest.fn()

        req.makeRequest({ connectionRetryCount: 43, logLevel: 'warn' }, 'some_id')
        expect((req['_request'] as jest.Mock).mock.calls[0][0].foo).toBe('bar')
        expect((req['_request'] as jest.Mock).mock.calls[0][0].sessionId).toBe('some_id')
        expect((req['_request'] as jest.Mock).mock.calls[0][2]).toBe(43)
    })

    it('should pick up the fullRequestOptions returned by transformRequest', async () => {
        const req = new WebDriverRequest('POST', '/foo/bar', { foo: 'bar' })
        const transformRequest = jest.fn().mockImplementation((requestOptions) => ({
            ...requestOptions,
            json: { foo: 'baz' }
        }))

        await req.makeRequest({
            transformRequest,
            protocol: 'https',
            hostname: 'localhost',
            port: 4445,
            path: '/wd/hub/',
            logLevel: 'warn'
        }, 'some_id')

        expect(got.mock.calls[0][1].json).toEqual({ foo: 'baz' })
    })

    it('should resolve with the body returned by transformResponse', async () => {
        const req = new WebDriverRequest('POST', 'session/:sessionId/element', { foo: 'requestBody' })

        const transformResponse = jest.fn().mockImplementation((response) => ({
            ...response,
            body: { value: { foo: 'transformedResponse' } },
        }))

        got.mockClear()
        const responseBody = await req.makeRequest({
            transformResponse,
            protocol: 'https',
            hostname: 'localhost',
            port: 4445,
            path: '/wd/hub/',
            logLevel: 'warn'
        }, 'foobar-123')

        expect(transformResponse.mock.calls[0][0]).toHaveProperty('body')
        expect(transformResponse.mock.calls[0][1].json).toEqual({ foo: 'requestBody' })
        await expect(responseBody).toEqual({ value: { foo: 'transformedResponse' } })
        got.mockClear()
    })

    describe('createOptions', () => {
        it('fails if command requires sessionId but none given', () => {
            const req = new WebDriverRequest('POST', `${path}/:sessionId/element`, {})
            expect(() => req['_createOptions']({ logLevel: 'warn' })).toThrow()
        })

        it('creates proper options set', () => {
            const req = new WebDriverRequest('POST', `${path}/:sessionId/element`, {})
            const options: gotOrig.Options = req['_createOptions']({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/',
                headers: { foo: 'bar' },
                connectionRetryTimeout: 10 * 1000,
                logLevel: 'warn'
            }, 'foobar12345')

            // @ts-ignore
            expect((options.agent! as gotOrig.Agents).https!.protocol).toBe('https:')
            // @ts-ignore
            expect((options.agent! as gotOrig.Agents).http!.protocol).toBe('http:')
            expect((options.url! as URL).href)
                .toBe('https://localhost:4445/session/foobar12345/element')
            expect(Object.keys(options.headers as Record<string, string>))
                .toEqual(['Content-Type', 'Connection', 'Accept', 'User-Agent', 'foo', 'Content-Length'])
            expect(options.timeout).toBe(10 * 1000)
        })

        it('passes a custom agent', () => {
            const req = new WebDriverRequest('POST', `${path}/:sessionId/element`, {})
            const agent = {
                http: new http.Agent({ keepAlive: true }),
                https: new https.Agent({ keepAlive: true })
            }
            const options = req['_createOptions']({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/',
                logLevel: 'warn',
                agent
            }, 'foobar12345')

            expect((options.agent! as gotOrig.Agents).http).toBe(agent.http)
            expect((options.agent! as gotOrig.Agents).https).toBe(agent.https)
        })

        it('ignors path when command is a hub command', () => {
            const req = new WebDriverRequest('POST', '/grid/api/hub', {}, true)
            const options = req['_createOptions']({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/',
                logLevel: 'warn'
            }, 'foobar12345')
            expect((options.url as URL).href).toBe('https://localhost:4445/grid/api/hub')
        })

        it('should add auth if user and key is given', () => {
            const req = new WebDriverRequest('POST', path, { some: 'body' })
            const options = req['_createOptions']({
                ...defaultOptions,
                user: 'foo',
                key: 'bar',
                path: '/',
                logLevel: 'warn'
            })
            expect(options.username).toEqual('foo')
            expect(options.password).toEqual('bar')
            expect(options.json).toEqual({ some: 'body' })
        })

        it('sets request body to "undefined" when request object is empty and DELETE is used', () => {
            const req = new WebDriverRequest('DELETE', path, {})
            const options = req['_createOptions']({
                ...defaultOptions,
                path: '/',
                logLevel: 'warn'
            })
            expect(Boolean(options.json)).toEqual(false)
        })

        it('sets request body to "undefined" when request object is empty and GET is used', () => {
            const req = new WebDriverRequest('GET', `${path}/title`, {})
            const options = req['_createOptions']({
                ...defaultOptions,
                path: '/',
                logLevel: 'warn'
            })
            expect(Boolean(options.json)).toEqual(false)
        })

        it('should attach an empty object body when POST is used', () => {
            const req = new WebDriverRequest('POST', '/status', {})
            const options = req['_createOptions']({
                ...defaultOptions,
                path: '/',
                logLevel: 'warn'
            })
            expect(options.json).toEqual({})
        })

        it('should add the Content-Length header when a request object has a body', () => {
            const req = new WebDriverRequest('POST', path, { foo: 'bar' })
            const options = req['_createOptions']({
                ...defaultOptions,
                path: '/',
                logLevel: 'warn'
            })
            expect(Object.keys(options.headers as Record<string, string>))
                .toEqual(['Content-Type', 'Connection', 'Accept', 'User-Agent', 'Content-Length'])
            expect((options.headers as Record<string, string>)['Content-Length']).toBe('13')
        })

        it('should add Content-Length as well any other header provided in the request options if there is body in the request object', () => {
            const req = new WebDriverRequest('POST', path, { foo: 'bar' })
            const options = req['_createOptions']({
                ...defaultOptions, path: '/',
                headers: { foo: 'bar' },
                logLevel: 'warn'
            })
            const headers = options.headers as Record<string, string>
            expect(Object.keys(headers)).toContain('Content-Length')
            expect(headers.foo).toContain('bar')
            expect(headers['Content-Length']).toBe('13')
        })

        it('should add only the headers provided if the request body is empty', () => {
            const req = new WebDriverRequest('POST', path)
            const options = req['_createOptions']({
                ...defaultOptions,
                path: '/',
                headers: { foo: 'bar' },
                logLevel: 'warn'
            })
            const headers = options.headers as Record<string, string>
            expect(Object.keys(headers)).not.toContain('Content-Length')
            expect(headers.foo).toContain('bar')
        })

        describe('rejectUnauthorized', () => {
            const defaults = {
                ...defaultOptions,
                path: '/',
                headers: { foo: 'bar' },
                logLevel: 'warn' as WebDriverLogTypes
            }

            beforeEach(function() {
                delete process.env.STRICT_SSL
                delete process.env.strict_ssl
            })

            it('should contain key "rejectUnauthorized" with value "false" when "strictSSL" argument is given with false', () => {
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions']({ ...defaults, strictSSL: false })

                expect(options.https?.rejectUnauthorized).toEqual(false)
            })

            it('should contain key "rejectUnauthorized" with value "false" when environment variable "STRICT_SSL" is defined with value "false"', () => {
                process.env['STRICT_SSL'] = 'false'
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(false)
            })

            it('should contain key "rejectUnauthorized" with value "false" when environment variable "strict_ssl" is defined with value "false"', () => {
                process.env['strict_ssl'] = 'false'
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(false)
            })

            it('should contain key "rejectUnauthorized" with value "true" when "strictSSL" argument is given with true', () => {
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions']({ ...defaults, strictSSL: true })
                expect(options.https?.rejectUnauthorized).toEqual(true)
            })

            it('should contain key "rejectUnauthorized" with value "true" when environment variable "STRICT_SSL" is defined with value "true"', () => {
                process.env['STRICT_SSL'] = 'true'
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(true)
            })

            it('should contain key "rejectUnauthorized" with value "true" when environment variable "strict_ssl" is defined with value "true"', () => {
                process.env['strict_ssl'] = 'true'
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(true)
            })

            it('should contain key "rejectUnauthorized" with value "true" when environment variable "STRICT_SSL" / "strict_ssl" is not defined', () => {
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(true)
            })

            it('should contain key "rejectUnauthorized" with value "true" when environment variable "STRICT_SSL" is defined with any other value than "false"', () => {
                process.env['STRICT_SSL'] = 'foo'
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(true)
            })

            it('should contain key "rejectUnauthorized" with value "true" when environment variable "strict_ssl" is defined with any other value than "false"', () => {
                process.env['strict_ssl'] = 'foo'
                const req = new WebDriverRequest('POST', path, {})
                const options = req['_createOptions'](defaults)
                expect(options.https?.rejectUnauthorized).toEqual(true)
            })
        })
    })

    describe('_request', () => {
        it('should make a request', async () => {
            const expectedResponse = { value: { 'element-6066-11e4-a52e-4f735466cecf': 'some-elem-123' } }
            const req = new WebDriverRequest('POST', path, {})
            req.emit = jest.fn()

            const opts = Object.assign(
                req.defaultOptions,
                { url: { pathname: '/session/foobar-123/element' } }
            )
            const res = await req['_request'](opts)

            expect(res).toEqual(expectedResponse)
            expect(req.emit).toBeCalledWith('response', { result: expectedResponse })
        })

        it('should short circuit if request throws a stale element exception', async () => {
            const req = new WebDriverRequest('POST', 'session/:sessionId/element', {})
            req.emit = jest.fn()

            const opts = Object.assign(req.defaultOptions, {
                url: { pathname: '/session/foobar-123/element/some-sub-sub-elem-231/click' },
                body: { foo: 'bar' }
            })

            const error = await req['_request'](opts).catch(err => err)
            expect(error.message).toBe('element is not attached to the page document')
            expect((req.emit as jest.Mock).mock.calls).toHaveLength(1)
            expect(warn.mock.calls).toHaveLength(1)
            expect(warn.mock.calls).toEqual([['Request encountered a stale element - terminating request']])
        })

        it('should not fail code due to an empty server response', async () => {
            const req = new WebDriverRequest('POST', path, {})
            req.emit = jest.fn()

            const opts = Object.assign(req.defaultOptions, { url: { pathname: '/empty' } })
            await expect(req['_request'](opts)).rejects.toEqual(new Error('Response has empty body'))
            expect((req.emit as jest.Mock).mock.calls).toHaveLength(1)
            expect(warn.mock.calls).toHaveLength(0)
            expect(error.mock.calls).toHaveLength(1)
        })

        it('should retry requests but still fail', async () => {
            const req = new WebDriverRequest('POST', path, {})
            req.emit = jest.fn()

            const opts = Object.assign(req.defaultOptions, { url: { pathname: '/failing' } })
            await expect(req['_request'](opts, undefined, 2)).rejects.toEqual(new Error('unknown error'))
            expect((req.emit as jest.Mock).mock.calls).toHaveLength(3)
            expect(warn.mock.calls).toHaveLength(2)
            expect(error.mock.calls).toHaveLength(1)
        })

        it('should retry and eventually respond', async () => {
            const req = new WebDriverRequest('POST', path, {})
            req.emit = jest.fn()

            const opts = Object.assign(req.defaultOptions, { url: { pathname: '/failing' }, json: { foo: 'bar' } })
            expect(await req['_request'](opts, undefined, 3)).toEqual({ value: 'caught' })
            expect((req.emit as jest.Mock).mock.calls).toHaveLength(4)
            expect(warn.mock.calls).toHaveLength(3)
            expect(error.mock.calls).toHaveLength(0)
        })

        it('should manage hub commands', async () => {
            const req = new WebDriverRequest('POST', '/grid/api/hub', {}, true)
            expect(await req.makeRequest({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/',
                logLevel: 'warn'
            }, 'foobar')).toEqual({ value: { some: 'config' } })
        })

        it('should fail if hub command is called on node', async () => {
            const req = new WebDriverRequest('POST', '/grid/api/testsession', {}, true)
            const result = await req.makeRequest({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/',
                logLevel: 'warn'
            }, 'foobar').then(
                (res) => res,
                (e) => e
            )
            expect(result.message).toBe('Command can only be called to a Selenium Hub')
        })

        it('should throw if timeout happens too often', async () => {
            const retryCnt = 3
            const req = new WebDriverRequest('POST', '/timeout', {}, true)
            const reqRetryCnt = jest.fn()
            req.on('retry', reqRetryCnt)
            const result = await req.makeRequest({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/timeout',
                connectionRetryCount: retryCnt,
                logLevel: 'warn'
            }, 'foobar').then(
                (res) => res,
                (e) => e
            )
            expect(result.code).toBe('ETIMEDOUT')
            expect(reqRetryCnt).toBeCalledTimes(retryCnt)
        })

        it('should return proper response if retry passes', async () => {
            const retryCnt = 7
            const req = new WebDriverRequest('POST', '/timeout', {}, true)
            const reqRetryCnt = jest.fn()
            req.on('retry', reqRetryCnt)
            const result = await req.makeRequest({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/timeout',
                connectionRetryCount: retryCnt,
                logLevel: 'warn'
            }, 'foobar').then(
                (res) => res,
                (e) => e
            )
            expect(result).toEqual({ value: { value: {} } })
            expect(reqRetryCnt).toBeCalledTimes(5)
        })

        it('should throw if request error is unknown', async () => {
            const req = new WebDriverRequest('POST', '/sumoerror', {}, true)
            const result = await req.makeRequest({
                protocol: 'https',
                hostname: 'localhost',
                port: 4445,
                path: '/sumoerror',
                connectionRetryCount: 0,
                logLevel: 'warn'
            }, 'foobar').then(
                (res) => res,
                (e) => e
            )
            expect(result.message).toBe('ups')
        })
    })

    afterEach(() => {
        // @ts-ignore
        got.retryCnt = 0

        got.mockClear()
        warn.mockClear()
        error.mockClear()
    })
})
