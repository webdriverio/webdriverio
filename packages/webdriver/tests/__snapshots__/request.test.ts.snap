// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webdriver request > defines correct exceptions for request retries 1`] = `
[
  "/session/:sessionId/actions",
]
`;
